File: app/src/main/java/com/caitlynwiley/pettracker/TrackerFragment.java
package com.caitlynwiley.pettracker;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;

import com.caitlynwiley.pettracker.models.Day;
import com.caitlynwiley.pettracker.models.Pet;
import com.caitlynwiley.pettracker.models.TrackerEvent;
import com.caitlynwiley.pettracker.models.TrackerItem;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Locale;
import java.util.TimeZone;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class TrackerFragment extends Fragment implements View.OnClickListener {

    private View mFragView;

    private Animation mRotateForward;
    private Animation mRotateBackward;
    private Animation mMiniAppear;
    private Animation mMiniDisappear;
    private Animation mLabelAppear;
    private Animation mLabelDisappear;

    private TextView mPottyFabLabel;
    private TextView mFeedFabLabel;
    private TextView mLetOutFabLabel;
    private FloatingActionButton mTrackerFab;
    private FloatingActionButton mPottyFab;
    private FloatingActionButton mFeedFab;
    private FloatingActionButton mLetOutFab;
    private EventAdapter mAdapter;

    private DatabaseReference mDatabase = FirebaseDatabase.getInstance().getReference();
    private ArrayList<String> pets = new ArrayList<>();
    private String mUID;
    private FirebaseUser mUser;
    private FirebaseAuth mAuth;

    private boolean mIsFabOpen;
    private boolean mPetsListWasEmpty = true;
    private Pet mPet;

    private ChildEventListener eventListener;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        mFragView = inflater.inflate(R.layout.tracker_fragment, container, false);
        mAuth = FirebaseAuth.getInstance();
        mUser = mAuth.getCurrentUser();
        mUID = mUser.getUid();
        //parentActivity = getActivity();
        mRotateForward = AnimationUtils.loadAnimation(getContext(), R.anim.fab_spin_forward);
        mRotateBackward = AnimationUtils.loadAnimation(getContext(), R.anim.fab_spin_backward);

        mMiniAppear = AnimationUtils.loadAnimation(getContext(), R.anim.mini_fab_appear);
        mMiniAppear.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {

            }

            @Override
            public void onAnimationEnd(Animation animation) {
                mPottyFab.setVisibility(View.VISIBLE);
                mLetOutFab.setVisibility(View.VISIBLE);
                mFeedFab.setVisibility(View.VISIBLE);
                mPottyFabLabel.setVisibility(View.VISIBLE);
                mLetOutFabLabel.setVisibility(View.VISIBLE);
                mFeedFabLabel.setVisibility(View.VISIBLE);
            }

            @Override
            public void onAnimationRepeat(Animation animation) {

            }
        });

        mMiniDisappear = AnimationUtils.loadAnimation(getContext(), R.anim.mini_fab_disappear);
        mMiniDisappear.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {

            }

            @Override
            public void onAnimationEnd(Animation animation) {
                mPottyFab.setVisibility(View.GONE);
                mLetOutFab.setVisibility(View.GONE);
                mFeedFab.setVisibility(View.GONE);
                mPottyFabLabel.setVisibility(View.GONE);
                mLetOutFabLabel.setVisibility(View.GONE);
                mFeedFabLabel.setVisibility(View.GONE);
            }

            @Override
            public void onAnimationRepeat(Animation animation) {

            }
        });

        mLabelAppear = AnimationUtils.loadAnimation(getContext(), R.anim.fab_label_appear);
        mLabelDisappear = AnimationUtils.loadAnimation(getContext(), R.anim.fab_label_disappear);

        mIsFabOpen = false;
        // get views
        mTrackerFab = mFragView.findViewById(R.id.tracker_fab);
        mPottyFab = mFragView.findViewById(R.id.track_potty_fab);
        mFeedFab = mFragView.findViewById(R.id.track_fed_fab);
        mLetOutFab = mFragView.findViewById(R.id.track_let_out_fab);
        mPottyFabLabel = mFragView.findViewById(R.id.potty_fab_label);
        mFeedFabLabel = mFragView.findViewById(R.id.fed_fab_label);
        mLetOutFabLabel = mFragView.findViewById(R.id.let_out_fab_label);

        // set on click listeners
        mTrackerFab.setOnClickListener(this);
        mPottyFab.setOnClickListener(this);
        mFeedFab.setOnClickListener(this);
        mLetOutFab.setOnClickListener(this);
        mPottyFabLabel.setOnClickListener(this);
        mFeedFabLabel.setOnClickListener(this);
        mLetOutFabLabel.setOnClickListener(this);

        RecyclerView recyclerView = mFragView.findViewById(R.id.tracker_items);

        // use a linear layout manager
        recyclerView.setLayoutManager(new LinearLayoutManager(mFragView.getContext()));

        // specify an adapter
        mAdapter = new EventAdapter(mFragView);
        recyclerView.setAdapter(mAdapter);

        // listens to changes made to user's mPet list
        mDatabase.child("users").child(mUID).child("pets").addChildEventListener(new ChildEventListener() {
            @Override
            public void onChildAdded(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {
                pets.add(dataSnapshot.getKey());
                if (mPetsListWasEmpty) {
                    mDatabase.child("pets").child(pets.get(0)).child("events").addChildEventListener(eventListener);
                    mPetsListWasEmpty = false;

                    mDatabase.child("pets").child(pets.get(0)).addValueEventListener(new ValueEventListener() {
                        @Override
                        public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                            mPet = dataSnapshot.getValue(Pet.class);
                        }

                        @Override
                        public void onCancelled(@NonNull DatabaseError databaseError) {

                        }
                    });
                }
            }

            @Override
            public void onChildChanged(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {

            }

            @Override
            public void onChildRemoved(@NonNull DataSnapshot dataSnapshot) {
                pets.remove(dataSnapshot.getKey());
                mAdapter.notifyDataSetChanged();
            }

            @Override
            public void onChildMoved(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {

            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });

        eventListener = new ChildEventListener() {

            @Override
            public void onChildAdded(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {
                TrackerItem item;
                if (dataSnapshot.child("itemType").getValue(String.class).equals("event")) {
                    item = dataSnapshot.getValue(TrackerEvent.class);
                    Log.d("fragment", "event");
                } else {
                    item = dataSnapshot.getValue(Day.class);
                    ((Day) item).setContext(getContext());
                    Log.d("fragment", "day");
                }
                if (mAdapter.getItemCount() == 0 && item instanceof TrackerEvent) {
                    // need to insert a day item before the event
                    TrackerEvent event = (TrackerEvent) item;
                    Day d = new Day(getContext(), event.getDate());
                    String id = mDatabase.child("pets").child(pets.get(0)).child("events").push().getKey();
                    d.setId(id);
                    Calendar c = Calendar.getInstance();
                    c.set(event.getYear(), event.getMonth() - 1, event.getDay(), 0, 0, 0);
                    c.set(Calendar.MILLISECOND, 0);
                    d.setUtcMillis(c.getTimeInMillis() - c.getTimeZone().getOffset(c.getTimeInMillis()));
                    mDatabase.child("pets").child(pets.get(0)).child("events").child(id).setValue(d);
                }
                mAdapter.addItem(item);
                mAdapter.notifyDataSetChanged();
            }

            @Override
            public void onChildChanged(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {

            }

            @Override
            public void onChildRemoved(@NonNull DataSnapshot dataSnapshot) {
                mAdapter.removeEvent(dataSnapshot.getValue(TrackerItem.class));
                mAdapter.notifyDataSetChanged();
            }

            @Override
            public void onChildMoved(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {

            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        };

        ItemTouchHelper itemTouchHelper = new ItemTouchHelper(new SwipeToDeleteHelper(mAdapter, getContext()));
        itemTouchHelper.attachToRecyclerView(recyclerView);

        return mFragView;
    }

    @Override
    public void onClick(View v) {
        int id = v.getId();
        switch (id) {
            case R.id.tracker_fab:
                if (mIsFabOpen) {
                    closeFab();
                } else {
                    openFab();
                }
                mIsFabOpen = !mIsFabOpen;
                break;
            case R.id.track_fed_fab:
                addEvent(TrackerEvent.EventType.FEED);
                break;
            case R.id.track_potty_fab:
                addEvent(TrackerEvent.EventType.POTTY);
                break;
            case R.id.track_let_out_fab:
                addEvent(TrackerEvent.EventType.WALK);
                break;
        }
    }

    private void addEvent(final TrackerEvent.EventType type) {

        switch (type) {
            case FEED:
                trackFeed();
                break;
            case WALK:
                trackWalk();
                break;
            case POTTY:
                trackPotty();
                break;
        }
    }

    private void trackPotty() {
        View v = getLayoutInflater().inflate(R.layout.track_potty_dialog, null);
        ((RadioButton) v.findViewById(R.id.pet0)).setText(mPet.getName());
        final Context context = this.getActivity().getApplicationContext();

        new AlertDialog.Builder(getContext())
                .setView(v)
                .setPositiveButton(R.string.save, (dialog, which) -> {
                    AlertDialog d = (AlertDialog) dialog;
                    // pet
                    RadioGroup radioGroup = d.findViewById(R.id.pet_radio_group);
                    int radioButtonID = radioGroup.getCheckedRadioButtonId();
                    View radioButton = radioGroup.findViewById(radioButtonID);
                    int petChosen = radioGroup.indexOfChild(radioButton);
                    String petId = pets.get(petChosen);
                    // type
                    boolean num1 = ((CheckBox) d.findViewById(R.id.number1)).isChecked();
                    boolean num2 = ((CheckBox) d.findViewById(R.id.number2)).isChecked();
                    //date
                    Calendar c = Calendar.getInstance();
                    String date = String.format(Locale.US, "%2d/%2d/%4d",
                            c.get(Calendar.MONTH) + 1, c.get(Calendar.DAY_OF_MONTH),
                            c.get(Calendar.YEAR));
                    if (!mAdapter.getMostRecentDate().equals(date)) {
                        addDayToList(context, c, petId, date);
                    }
                    String id = mDatabase.child("pets").child(petId).child("events").push().getKey();
                    TrackerEvent e = new TrackerEvent.Builder(TrackerEvent.EventType.POTTY)
                            .setDate(date)
                            .setMillis(c.getTimeInMillis() - c.getTimeZone().getOffset(c.getTimeInMillis()))
                            .setNumber1(num1)
                            .setNumber2(num2)
                            .setPetId(petId)
                            .setId(id)
                            .build();
                    mDatabase.child("pets").child(petId).child("events").child(id).setValue(e);
                })
                .setNegativeButton(R.string.cancel, (dialog, which) -> dialog.cancel())
                .create()
                .show();
    }

    private void trackWalk() {
        View v = getLayoutInflater().inflate(R.layout.track_walk_dialog, null);
        ((RadioButton) v.findViewById(R.id.pet0)).setText(mPet.getName());
        final Context context = this.getActivity().getApplicationContext();

        new AlertDialog.Builder(getContext())
                .setView(v)
                .setPositiveButton(R.string.save, (dialog, which) -> {
                    AlertDialog d = (AlertDialog) dialog;
                    // pet
                    RadioGroup radioGroup = d.findViewById(R.id.pet_radio_group);
                    int radioButtonID = radioGroup.getCheckedRadioButtonId();
                    View radioButton = radioGroup.findViewById(radioButtonID);
                    int petChosen = radioGroup.indexOfChild(radioButton);
                    String petId = pets.get(petChosen);
                    // duration
                    int hours = Integer.parseInt(((EditText) d.findViewById(R.id.walk_duration_hours))
                            .getText().toString());
                    int mins = Integer.parseInt(((EditText) d.findViewById(R.id.walk_duration_mins))
                            .getText().toString());
                    //date
                    Calendar c = Calendar.getInstance();
                    String date = String.format(Locale.US, "%2d/%2d/%4d",
                            c.get(Calendar.MONTH) + 1, c.get(Calendar.DAY_OF_MONTH),
                            c.get(Calendar.YEAR));
                    if (!mAdapter.getMostRecentDate().equals(date)) {
                        addDayToList(context, c, petId, date);
                    }
                    String id = mDatabase.child("pets").child(petId).child("events").push().getKey();
                    TrackerEvent e = new TrackerEvent.Builder(TrackerEvent.EventType.WALK)
                            .setDate(date)
                            .setMillis(c.getTimeInMillis() - c.getTimeZone().getOffset(c.getTimeInMillis()))
                            .setWalkLength(hours, mins)
                            .setPetId(petId)
                            .setId(id)
                            .build();
                    mDatabase.child("pets").child(petId).child("events").child(id).setValue(e);
                })
                .setNegativeButton(R.string.cancel, (dialog, which) -> dialog.cancel())
                .create()
                .show();
    }

    private void trackFeed() {
        View v = getLayoutInflater().inflate(R.layout.track_meal_dialog, null);
        ((RadioButton) v.findViewById(R.id.pet0)).setText(mPet.getName());
        final Context context = this.getActivity().getApplicationContext();

        new AlertDialog.Builder(getContext())
                .setView(v)
                .setPositiveButton(R.string.save, (dialog, which) -> {
                    AlertDialog d = (AlertDialog) dialog;
                    // pet
                    RadioGroup radioGroup = d.findViewById(R.id.pet_radio_group);
                    int radioButtonID = radioGroup.getCheckedRadioButtonId();
                    View radioButton = radioGroup.findViewById(radioButtonID);
                    int petChosen = radioGroup.indexOfChild(radioButton);
                    String petId = pets.get(petChosen);
                    // amount
                    double cupsFood = Double.parseDouble(((EditText) d.findViewById(R.id.num_cups))
                            .getText().toString());
                    //date
                    Calendar c = Calendar.getInstance();
                    String date = String.format(Locale.US, "%2d/%2d/%4d",
                            c.get(Calendar.MONTH) + 1, c.get(Calendar.DAY_OF_MONTH),
                            c.get(Calendar.YEAR));
                    if (!mAdapter.getMostRecentDate().equals(date)) {
                        addDayToList(context, c, petId, date);
                    }
                    String id = mDatabase.child("pets").child(petId).child("events").push().getKey();
                    TrackerEvent e = new TrackerEvent.Builder(TrackerEvent.EventType.FEED)
                            .setDate(date)
                            .setMillis(c.getTimeInMillis() - c.getTimeZone().getOffset(c.getTimeInMillis()))
                            .setCupsFood(cupsFood)
                            .setPetId(petId)
                            .setId(id)
                            .build();
                    mDatabase.child("pets").child(petId).child("events").child(id).setValue(e);
                })
                .setNegativeButton(R.string.cancel, (dialog, which) -> dialog.cancel())
                .create()
                .show();
    }

    private void addDayToList(Context context, Calendar c, String petId, String date) {
        Day day = new Day(context, date);
        String dayId = mDatabase.child("pets").child(petId).child("events").push().getKey();
        day.setId(dayId);
        c.set(Calendar.HOUR_OF_DAY, 0);
        c.set(Calendar.MINUTE, 0);
        c.set(Calendar.SECOND, 0);
        c.set(Calendar.MILLISECOND, 0);
        day.setUtcMillis(c.getTimeInMillis());
        mDatabase.child("pets").child(petId).child("events").child(dayId).setValue(day);
    }

    private void openFab() {
        mTrackerFab.startAnimation(mRotateForward);
        mLetOutFab.startAnimation(mMiniAppear);
        mPottyFab.startAnimation(mMiniAppear);
        mFeedFab.startAnimation(mMiniAppear);
        mPottyFabLabel.startAnimation(mLabelAppear);
        mFeedFabLabel.startAnimation(mLabelAppear);
        mLetOutFabLabel.startAnimation(mLabelAppear);
    }

    private void closeFab() {
        mTrackerFab.startAnimation(mRotateBackward);
        mLetOutFab.startAnimation(mMiniDisappear);
        mPottyFab.startAnimation(mMiniDisappear);
        mFeedFab.startAnimation(mMiniDisappear);
        mPottyFabLabel.startAnimation(mLabelDisappear);
        mFeedFabLabel.startAnimation(mLabelDisappear);
        mLetOutFabLabel.startAnimation(mLabelDisappear);
    }
}

File: app/src/main/java/com/caitlynwiley/pettracker/views/fragments/TrackerFragment.kt
package com.caitlynwiley.pettracker.views.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.compose.animation.ExperimentalAnimationApi
import androidx.compose.material.Surface
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.ui.platform.ComposeView
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import com.caitlynwiley.pettracker.EventAdapter
import com.caitlynwiley.pettracker.models.TrackerItem
import com.caitlynwiley.pettracker.repository.PetTrackerRepository
import com.caitlynwiley.pettracker.theme.PetTrackerTheme
import com.caitlynwiley.pettracker.viewmodel.TrackerViewModel
import com.caitlynwiley.pettracker.viewmodel.TrackerViewModelFactory
import com.google.accompanist.swiperefresh.SwipeRefresh
import com.google.accompanist.swiperefresh.rememberSwipeRefreshState
import org.apache.commons.text.RandomStringGenerator
import java.util.*

class TrackerFragment : Fragment() {
    private val viewModel by viewModels<TrackerViewModel>(
        factoryProducer = { TrackerViewModelFactory(PetTrackerRepository(), "") }
    )

    private lateinit var mAdapter: EventAdapter

    private val randomStringGenerator: RandomStringGenerator =
        RandomStringGenerator.Builder().withinRange(charArrayOf('0', '9'), charArrayOf('a', 'z'),
            charArrayOf('A', 'Z')).build()

    @ExperimentalAnimationApi
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        return ComposeView(requireContext()).apply {
            setContent {
                PetTrackerTheme {
                    Surface {
                        val trackerItems by viewModel.trackerItems.observeAsState(listOf())
                        val isRefreshing by viewModel.isRefreshing.collectAsState()

                        SwipeRefresh(state = rememberSwipeRefreshState(isRefreshing = isRefreshing),
                            onRefresh = { viewModel.refresh() }
                        ) {
                            TrackerScreen(items = trackerItems)
                        }
                    }
                }
            }
        }

//        mRotateForward = AnimationUtils.loadAnimation(context, R.anim.fab_spin_forward)
//        mRotateBackward = AnimationUtils.loadAnimation(context, R.anim.fab_spin_backward)
//
//        // setup swipe-to-delete and undo logic
//        val itemTouchHelper = ItemTouchHelper(SwipeToDeleteHelper(mAdapter, requireContext()))
//        itemTouchHelper.attachToRecyclerView(recyclerView)
    }

    /*
    Positive btn: "Save"
    onClick (pseudocode):
        // type
        val num1 = peeCheckBox.isChecked
        val num2 = poopCheckBox.isChecked
        //date
        val c = Calendar.getInstance()
        val date = String.format(
            Locale.US, "%02d/%02d/%4d",
            c[Calendar.MONTH] + 1, c[Calendar.DAY_OF_MONTH],
            c[Calendar.YEAR]
        )
        if (mAdapter.mostRecentDate != date) {
            addDayToList(c, viewModel.pet.value?.id ?: "")
        }
        val e = TrackerItem.Builder()
            .setEventType(TrackerItem.EventType.POTTY)
            .setMillis(System.currentTimeMillis())
            .setNumber1(num1)
            .setNumber2(num2)
            .setPetId(viewModel.pet.value?.id ?: "")
            .setItemType("event")
            .setId(randomStringGenerator.generate(30))
            .build()
        viewModel.addTrackerItem(e)

    Negative btn: "Cancel"
    onClick: dialog.cancel()
     */

    /*
    Positive btn: "Save"
    onClick (pseudocode):
        val hours = hoursTextField.getValue
        val mins = minutesTextField.getValue
        //date
        val c = Calendar.getInstance()
        val date = String.format(
            Locale.US, "%02d/%02d/%4d",
            c[Calendar.MONTH] + 1, c[Calendar.DAY_OF_MONTH],
            c[Calendar.YEAR]
        )
        if (mAdapter.mostRecentDate != date) {
            addDayToList(c, viewModel.pet.value?.id ?: "")
        }
        val e = TrackerItem.Builder()
            .setEventType(TrackerItem.EventType.WALK)
            .setMillis(System.currentTimeMillis())
            .setWalkLength(hours, mins)
            .setPetId(viewModel.pet.value?.id ?: "")
            .setItemType("event")
            .setId(randomStringGenerator.generate(30))
            .build()
        viewModel.addTrackerItem(e)

    Negative btn: "Cancel"
    onClick: dialog.cancel()
     */

    /*
    Positive btn: "Save"
    onClick (pseudocode):
        if (cups.isEmpty() || cups.getValue == "0") {
            // stay open and show error message
        } else {
            // amount
            val cupsFood = cups.getValue
            //date
            val c = Calendar.getInstance()
            val date = String.format(
                Locale.US, "%02d/%02d/%4d",
                c[Calendar.MONTH] + 1, c[Calendar.DAY_OF_MONTH],
                c[Calendar.YEAR]
            )
            if (mAdapter.mostRecentDate != date) {
                addDayToList(c, viewModel.pet.value?.id ?: "")
            }
            val e = TrackerItem.Builder()
                .setEventType(TrackerItem.EventType.FEED)
                .setMillis(System.currentTimeMillis())
                .setCupsFood(cupsFood)
                .setPetId(viewModel.pet.value?.id ?: "")
                .setItemType("event")
                .setId(randomStringGenerator.generate(30))
                .build()
            viewModel.addTrackerItem(e)
        }

    Negative btn: "Cancel"
    onClick: dialog.cancel()
     */

    private fun addDayToList(c: Calendar, petId: String) {
        val day = TrackerItem.Builder()
            .setMillis(System.currentTimeMillis())
            .setItemType("day")
            .setPetId(petId)
            .setId(randomStringGenerator.generate(30))
            .build()
        val now = System.currentTimeMillis()
        c.timeInMillis = now
        c[Calendar.HOUR] = 0
        c[Calendar.MINUTE] = 0
        c[Calendar.SECOND] = 0
        c[Calendar.MILLISECOND] = 0
        c.add(Calendar.MILLISECOND, viewModel.gmtTimeZone.getOffset(now))
        day.setUtcMillis(c.timeInMillis)
        viewModel.addTrackerItem(day)
    }
}